File: java/pig-latin/src/main/java/PigLatin.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class PigLatin {

  private static final Character[] VOWELS =
    new Character[] {'a', 'e', 'i', 'o', 'u'};
  private static final Set<Character> VOWELS_SET =
    new HashSet<Character>(Arrays.asList(VOWELS));

  public static String translate(String phrase) {
    String[] words = phrase.split(" ");
    for (int i = 0; i < words.length; i++) {
      words[i] = translateWord(words[i]);
    }
    return String.join(" ", words);
  }

  private static String translateWord(String word) {
    if (singleVowel(word)){
      return word + "ay";
    } else if (tripleConsonant(word)) {
      return word.substring(3) + word.substring(0, 3) + "ay";
    } else if (doubleConsonant(word)) {
      return word.substring(2) + word.substring(0, 2) + "ay";
    } else {
      return word.substring(1) + word.substring(0, 1) + "ay";
    }
  }

  private static boolean singleVowel(String word) {
    return ((VOWELS_SET.contains(word.charAt(0))) ||
            (word.startsWith("xr")) ||
            (word.startsWith("yt")));
  }

  private static boolean tripleConsonant(String word) {
    return (word.startsWith("squ") ||
            word.startsWith("sch") ||
            word.startsWith("thr"));
  }

  private static boolean doubleConsonant(String word) {
    return (word.startsWith("ch") ||
            word.startsWith("qu") ||
            word.startsWith("th"));
  }
}

File: go/pig-latin/pig_latin.go
package piglatin

import (
	"fmt"
	"strings"
)

var vowels = []string{"a", "e", "i", "o", "u"}
var consonantSounds = []string{"squ", "thr", "th", "sch", "qu", "ch", "rh", "p", "k", "x", "q", "y", "m", "f", "r"}

func Sentence(sentence string) string {
	words := strings.Split(sentence, " ")
	translated := []string{}
	for _, word := range words {
		translated = append(translated, translate(word))
	}
	return strings.Join(translated, " ")
}

func translate(word string) string {
	if startsWithVowel(word) ||
		strings.HasPrefix(word, "xr") ||
		strings.HasPrefix(word, "yt") {
		return handleVowel(word)
	}
	if startsWithConsonantSound(word) {
		return handleConsonantSound(word)
	}
	panic(fmt.Sprintf("unhandled word %v", word))
}

func startsWithVowel(sentence string) bool {
	for _, vowel := range vowels {
		if strings.HasPrefix(sentence, vowel) {
			return true
		}
	}
	return false
}

func startsWithConsonantSound(sentence string) bool {
	for _, cs := range consonantSounds {
		if strings.HasPrefix(sentence, cs) {
			return true
		}
	}
	return false
}

func handleConsonantSound(sentence string) string {
	for _, cs := range consonantSounds {
		if strings.HasPrefix(sentence, cs) {
			return strings.TrimPrefix(sentence, cs) + cs + "ay"
		}
	}
	panic(fmt.Sprintf("could not find consonant sound prefix for sentence %v", sentence))
}

func handleVowel(sentence string) string {
	return sentence + "ay"
}
